### PR TITLE
opt: optimize FuncDepSet.inClosureOf

### PR DESCRIPTION
**Pull request by mgartner**
#### opt/bench: add slow-query-5

This commit adds the new `slow-query-5` benchmark which tests a 40+ -way
join modelled off of a real-world production query. The query stresses
the efficiency of the optimizer, particularly in building and
manipulating thousands of functional dependencies for the thousands of
possible join orderings.

Release note: None

#### opt: optimize FuncDepSet.inClosureOf

`FuncDepSet.inClosureOf` is used frequently to manipulate a
`FuncDepSet`. In queries with many tables, columns, and joins, it can
contribute significantly to the latency of query planning, due to the
`O(n²)` time complexity of calculating a transitive closure, where `n`
is the number of dependencies in the set, and the relatively expensive
`SubsetOf` set operation.

This commit optimizes the function by avoiding some unnecessary
`SubsetOf` invocations in favor of the cheaper `Intersects` set
operator.

Epic: None

Release note (performance improvement): Query planning time has been
reduced significantly for some queries in which many tables are joined.

### CODE REVIEW

**Comment by mgartner** [comment]
I'll add benchmark results shortly.

**Comment by mgartner** [comment]
```
name                         old time/op    new time/op    delta
SlowQueries/slow-query-1-24    29.3ms ± 2%    28.7ms ± 2%   -2.12%  (p=0.032 n=5+5)
SlowQueries/slow-query-2-24     413ms ± 1%     406ms ± 1%   -1.66%  (p=0.008 n=5+5)
SlowQueries/slow-query-3-24    97.0ms ± 0%    96.7ms ± 1%     ~     (p=0.222 n=5+5)
SlowQueries/slow-query-4-24     687ms ± 0%     696ms ± 1%   +1.27%  (p=0.008 n=5+5)
SlowQueries/slow-query-5-24     4.34s ± 0%     2.91s ± 0%  -32.94%  (p=0.008 n=5+5)

name                         old alloc/op   new alloc/op   delta
SlowQueries/slow-query-1-24    6.83MB ± 0%    6.83MB ± 0%     ~     (p=0.222 n=5+5)
SlowQueries/slow-query-2-24    48.6MB ± 0%    48.5MB ± 0%     ~     (p=0.151 n=5+5)
SlowQueries/slow-query-3-24    44.8MB ± 0%    44.8MB ± 0%     ~     (p=0.222 n=5+5)
SlowQueries/slow-query-4-24     153MB ± 0%     153MB ± 0%     ~     (p=0.151 n=5+5)
SlowQueries/slow-query-5-24     405MB ± 0%     405MB ± 0%     ~     (p=1.000 n=4+4)

name                         old allocs/op  new allocs/op  delta
SlowQueries/slow-query-1-24     70.2k ± 0%     70.2k ± 0%     ~     (p=0.397 n=5+5)
SlowQueries/slow-query-2-24      391k ± 0%      391k ± 0%     ~     (p=0.500 n=5+5)
SlowQueries/slow-query-3-24      389k ± 0%      389k ± 0%     ~     (p=0.095 n=5+5)
SlowQueries/slow-query-4-24     3.23M ± 0%     3.23M ± 0%     ~     (p=0.690 n=5+5)
SlowQueries/slow-query-5-24     8.94M ± 0%     8.94M ± 0%     ~     (p=1.000 n=5+5)
```

```
name                                                    old time/op    new time/op    delta
Phases/kv-read/Simple/Explore-24                          49.7µs ± 0%    50.3µs ± 0%  +1.08%  (p=0.008 n=5+5)
Phases/kv-read-const/Simple/Explore-24                    50.4µs ± 0%    50.5µs ± 1%    ~     (p=0.222 n=5+5)
Phases/tpcc-new-order/Simple/Explore-24                    134µs ± 0%     134µs ± 0%    ~     (p=1.000 n=5+5)
Phases/tpcc-delivery/Simple/Explore-24                     115µs ± 1%     114µs ± 0%    ~     (p=0.095 n=5+5)
Phases/tpcc-delivery/Prepared/Explore-24                  43.2µs ± 1%    43.2µs ± 0%    ~     (p=0.690 n=5+5)
Phases/tpcc-stock-level/Simple/Explore-24                  608µs ± 0%     607µs ± 0%    ~     (p=0.421 n=5+5)
Phases/tpcc-stock-level/Prepared/Explore-24                263µs ± 0%     262µs ± 0%  -0.39%  (p=0.016 n=5+5)
Phases/many-columns-and-indexes-a/Simple/Explore-24        455µs ± 0%     452µs ± 0%  -0.75%  (p=0.008 n=5+5)
Phases/many-columns-and-indexes-a/Prepared/Explore-24      403µs ± 0%     400µs ± 0%  -0.66%  (p=0.008 n=5+5)
Phases/many-columns-and-indexes-b/Simple/Explore-24        544µs ± 0%     540µs ± 0%  -0.78%  (p=0.008 n=5+5)
Phases/many-columns-and-indexes-b/Prepared/Explore-24      458µs ± 0%     455µs ± 0%  -0.63%  (p=0.032 n=5+5)
Phases/many-columns-and-indexes-c/Simple/Explore-24        335ms ± 1%     326ms ± 1%  -2.63%  (p=0.008 n=5+5)
Phases/many-columns-and-indexes-c/Prepared/Explore-24      341ms ± 1%     328ms ± 2%  -3.70%  (p=0.008 n=5+5)
Phases/single-col-histogram-range/Simple/Explore-24        138µs ± 0%     137µs ± 0%    ~     (p=0.421 n=5+5)
Phases/single-col-histogram-range/Prepared/Explore-24     58.9µs ± 0%    58.7µs ± 0%    ~     (p=0.222 n=5+5)
Phases/batch-insert-one/Simple/Explore-24                  119µs ± 1%     120µs ± 1%  +1.40%  (p=0.032 n=5+5)
Phases/batch-insert-many/Simple/Explore-24                8.61ms ± 1%    8.64ms ± 1%    ~     (p=0.222 n=5+5)
Phases/ored-preds-100/Simple/Explore-24                   3.68ms ± 0%    3.67ms ± 0%  -0.31%  (p=0.016 n=5+5)
Phases/ored-preds-using-params-100/Simple/Explore-24      3.51ms ± 1%    3.48ms ± 0%  -1.03%  (p=0.008 n=5+5)
Phases/ored-preds-using-params-100/Prepared/Explore-24    1.56ms ± 1%    1.53ms ± 1%  -1.98%  (p=0.008 n=5+5)

name                                                    old alloc/op   new alloc/op   delta
Phases/kv-read/Simple/Explore-24                          15.4kB ± 0%    15.4kB ± 0%    ~     (p=0.167 n=5+5)
Phases/kv-read-const/Simple/Explore-24                    15.5kB ± 0%    15.5kB ± 0%    ~     (all equal)
Phases/tpcc-new-order/Simple/Explore-24                   32.7kB ± 0%    32.7kB ± 0%    ~     (p=0.405 n=5+5)
Phases/tpcc-delivery/Simple/Explore-24                    26.7kB ± 0%    26.7kB ± 0%    ~     (p=0.429 n=5+5)
Phases/tpcc-delivery/Prepared/Explore-24                  12.8kB ± 0%    12.8kB ± 0%    ~     (p=0.992 n=5+5)
Phases/tpcc-stock-level/Simple/Explore-24                  130kB ± 0%     130kB ± 0%    ~     (p=0.841 n=5+5)
Phases/tpcc-stock-level/Prepared/Explore-24               58.9kB ± 0%    58.9kB ± 0%    ~     (p=0.841 n=5+5)
Phases/many-columns-and-indexes-a/Simple/Explore-24       39.7kB ± 0%    39.7kB ± 0%    ~     (p=0.548 n=5+5)
Phases/many-columns-and-indexes-a/Prepared/Explore-24     5.50kB ± 0%    5.50kB ± 0%    ~     (all equal)
Phases/many-columns-and-indexes-b/Simple/Explore-24       48.0kB ± 0%    48.0kB ± 0%    ~     (p=0.444 n=5+5)
Phases/many-columns-and-indexes-b/Prepared/Explore-24     8.02kB ± 0%    8.02kB ± 0%    ~     (p=0.397 n=5+5)
Phases/many-columns-and-indexes-c/Simple/Explore-24        110MB ± 0%     110MB ± 0%    ~     (p=0.056 n=5+5)
Phases/many-columns-and-indexes-c/Prepared/Explore-24      109MB ± 0%     109MB ± 0%    ~     (p=0.222 n=5+5)
Phases/single-col-histogram-range/Simple/Explore-24       29.8kB ± 0%    29.8kB ± 0%    ~     (p=0.333 n=5+5)
Phases/single-col-histogram-range/Prepared/Explore-24     13.0kB ± 0%    13.0kB ± 0%    ~     (all equal)
Phases/batch-insert-one/Simple/Explore-24                 64.8kB ± 0%    64.8kB ± 0%    ~     (p=0.802 n=5+5)
Phases/batch-insert-many/Simple/Explore-24                3.92MB ± 0%    3.92MB ± 0%    ~     (p=0.548 n=5+5)
Phases/ored-preds-100/Simple/Explore-24                    945kB ± 0%     945kB ± 0%    ~     (p=0.421 n=5+5)
Phases/ored-preds-using-params-100/Simple/Explore-24       926kB ± 0%     926kB ± 0%    ~     (p=0.167 n=5+5)
Phases/ored-preds-using-params-100/Prepared/Explore-24     736kB ± 0%     736kB ± 0%    ~     (p=0.841 n=5+5)

name                                                    old allocs/op  new allocs/op  delta
Phases/kv-read/Simple/Explore-24                            97.0 ± 0%      97.0 ± 0%    ~     (all equal)
Phases/kv-read-const/Simple/Explore-24                      97.0 ± 0%      97.0 ± 0%    ~     (all equal)
Phases/tpcc-new-order/Simple/Explore-24                      177 ± 0%       177 ± 0%    ~     (all equal)
Phases/tpcc-delivery/Simple/Explore-24                       167 ± 0%       167 ± 0%    ~     (all equal)
Phases/tpcc-delivery/Prepared/Explore-24                    75.0 ± 0%      75.0 ± 0%    ~     (all equal)
Phases/tpcc-stock-level/Simple/Explore-24                    711 ± 0%       711 ± 0%    ~     (all equal)
Phases/tpcc-stock-level/Prepared/Explore-24                  341 ± 0%       341 ± 0%    ~     (all equal)
Phases/many-columns-and-indexes-a/Simple/Explore-24         83.0 ± 0%      83.0 ± 0%    ~     (all equal)
Phases/many-columns-and-indexes-a/Prepared/Explore-24       31.0 ± 0%      31.0 ± 0%    ~     (all equal)
Phases/many-columns-and-indexes-b/Simple/Explore-24          138 ± 0%       138 ± 0%    ~     (all equal)
Phases/many-columns-and-indexes-b/Prepared/Explore-24       49.0 ± 0%      49.0 ± 0%    ~     (all equal)
Phases/many-columns-and-indexes-c/Simple/Explore-24         762k ± 0%      762k ± 0%    ~     (p=0.310 n=5+5)
Phases/many-columns-and-indexes-c/Prepared/Explore-24       756k ± 0%      756k ± 0%    ~     (p=0.103 n=5+5)
Phases/single-col-histogram-range/Simple/Explore-24          524 ± 0%       524 ± 0%    ~     (all equal)
Phases/single-col-histogram-range/Prepared/Explore-24        251 ± 0%       251 ± 0%    ~     (all equal)
Phases/batch-insert-one/Simple/Explore-24                    325 ± 0%       325 ± 0%    ~     (all equal)
Phases/batch-insert-many/Simple/Explore-24                 28.5k ± 0%     28.5k ± 0%    ~     (p=0.881 n=5+5)
Phases/ored-preds-100/Simple/Explore-24                    4.53k ± 0%     4.53k ± 0%    ~     (p=0.333 n=4+5)
Phases/ored-preds-using-params-100/Simple/Explore-24       4.53k ± 0%     4.53k ± 0%    ~     (p=1.000 n=5+5)
Phases/ored-preds-using-params-100/Prepared/Explore-24     2.88k ± 0%     2.88k ± 0%    ~     (all equal)
```

Benchmarks were run on my GCE worker.

**Comment by mgartner** [comment]
I have an alternate approach in #113239 that reduces planning time for `slow-query-5` more than this PR, at the cost of significantly more memory usage (+40%). We should merge only one of them. Is the additional memory usage worth it? Also see my note here: https://github.com/cockroachdb/cockroach/pull/113239#issuecomment-1783425520

File: pkg/sql/opt/props/func_dep.go [hunk 1]
```diff
@@ -1761,10 +1761,27 @@ func (f *FuncDepSet) inClosureOf(cols, in opt.ColSet, strict bool) bool {
...
+		}
+
+		previouslySeen := i <= restartedAt
+		if previouslySeen && !fd.from.Intersects(recentlyAdded) {
+			// If the FD was seen in the last iteration before restarting and
+			// the columns most recently added to "in" do not intersect the
+			// "from" set, then the FD will not grow the closure, so it can be
+			// skipped. This is an optimization using the cheaper
+			// set-intersection operation (compare to the set-subset operations
+			// below), which is unnecessary for correctness.
```
Comment (mgartner) [inline pkg/sql/opt/props/func_dep.go:1780]: I realized this explanation isn't accurate. I'll rework it.

**Comment by DrewKimball** [comment]
Nice improvement! Seems like it's all upside. It looks like most of these calls are coming through `ReduceCols` - do you think we could exploit that for further gains?

**Review by rharding6373 — APPROVED** [review]
I'm in favor of this approach over the other approach that increases memory usage. I don't think it's worth the extra 8% performance improvement for one kind of query is worth the memory increases for most other kinds of slow queries we benchmark.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113237#-:-Ni0jtxF7GbR745YYX9C:b-d4s4aw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

**Comment by mgartner** [comment]
> Nice improvement! Seems like it's all upside. It looks like most of these calls are coming through ReduceCols - do you think we could exploit that for further gains?

Great idea. #113319 looks more promising than this PR. I'll close this for now.

**Pull request closed without merging**